src/Client/Updater.php

```diff
@@ -428,7 +428,21 @@ private function searchDelegatedRolesForTarget(TargetsMetadata $targetsMetadata,
         foreach ($targetsMetadata->getDelegatedKeys() as $keyId => $delegatedKey) {
             $this->signatureVerifier->addKey($keyId, $delegatedKey);
         }
+
+        $delegatedRoles = [];
         foreach ($targetsMetadata->getDelegatedRoles() as $delegatedRole) {
+            // Targets must match the paths of all roles in the delegation chain, so if the path does not match,
+            // do not evaluate this role or any roles it delegates to.
+            if ($delegatedRole->matchesPath($target)) {
+                $delegatedRoles[] = $delegatedRole;
+
+                if ($delegatedRole->isTerminating()) {
+                    break;
+                }
+            }
+        }
+
+        foreach ($delegatedRoles as $delegatedRole) {
             $delegatedRoleName = $delegatedRole->getName();
             if (in_array($delegatedRoleName, $searchedRoles, true)) {
```

Comment:
should we move this to loop above. all tests still pass

Reply:
We could, but I think we might have to change it later when I convert this to a set of promises, rather than an iterative loop.